!!! FEATURE: Configurable Fusion node type prototype generators

mficzel opened the pull request:
- If `fusion.prototypeGenerator` is set to a className the class has to implement the interface
  `\TYPO3\Neos\Domain\Service\DefaultPrototypeGeneratorInterface` and is used to generate the prototype-code
  for this node.
- If `fusion.prototypeGenerator` is set to `null` no prototype is created for this type.

For nodeTypes based on `TYPO3.Neos:Node`, `TYPO3.Neos:Document` or `TYPO3.Neos:Content` a backwards compatible implementation is added to the configuration and a new prototype generator for `TYPO3.Neos:Plugin` is added.

kitsunet commented:
I like. Would also allow us to add a plugin generator as that is always wrong :)

mficzel commented:
@kitsunet ... i'm impressed, you were almost faster then the mention bot.

What exactly should be added to the defaultPlugin Prototypes? I do'nt use it that often.

aertmann commented on TYPO3.Neos/Documentation/CreatingASite/NodeTypes/NodeTypeDefinition.rst:
```diff
@@ -119,6 +119,36 @@ The following options are allowed:
+        templatePath = 'resource://Vendor.Site/Private/Templates/NodeTypes/Content.SpecialNodeType.html'
+
+        # all properties of the nodeType are passed to the template
+        date =  ${q(node).property('date)}
```
double space

bwaidelich replied:
..and missing closing single quote

aertmann commented on TYPO3.Neos/Documentation/CreatingASite/NodeTypes/NodeTypeDefinition.rst:
```diff
@@ -119,6 +119,36 @@ The following options are allowed:
+
+    By default Neos has generators for all nodes of type TYPO3.Neos:Node and creates protoypes based on
+    TYPO3.TypoScript:Template. A template path is assumed based on the package-prefix and the nodetype-name. All properties
+    of the node are passed to the template. For nodeTypes of type TYPO3.Neos:Document or TYPO3.Neos:Content the
```
nodetype names should be quoted..

aertmann commented on TYPO3.Neos/Documentation/CreatingASite/NodeTypes/NodeTypeDefinition.rst:
```diff
@@ -119,6 +119,36 @@ The following options are allowed:
+    If ``options.fusionPrototypeGenerator`` is set to ``null`` no prototype is created for this type.
+
+    By default Neos has generators for all nodes of type TYPO3.Neos:Node and creates protoypes based on
+    TYPO3.TypoScript:Template. A template path is assumed based on the package-prefix and the nodetype-name. All properties
```
nodetype names should be quoted with ``

aertmann commented on TYPO3.Neos/Configuration/NodeTypes.yaml:
```diff
@@ -2,6 +2,8 @@
   label: "${String.cropAtWord(String.trim(String.stripTags(String.pregReplace(q(node).property('title') || q(node).property('text') || ((I18n.translate(node.nodeType.label) || node.nodeType.name) + (node.autoCreated ? ' (' + node.name + ')' : '')), '/<br\\W*?\\/?>|\\x{00a0}|[[^:print:]]|\\s+/u', ' '))), 100, '...')}"
   abstract: TRUE
+  options:
+    fusionPrototypeGenerator: TYPO3\Neos\Domain\Service\DefaultPrototypeGenerator
```
I'm debating if the key `options` is a good idea here.. options is pretty broad and all of this configuration is somehow options.. would a key like `fusion` and then prototypeGenerator on next level make more sense?

aertmann commented:
btw. what to do about the existing interface overriding? that would stop working as expected right? if so this is a breaking change..

aertmann commented:
@skurfuerst @bwaidelich maybe you guys have input to this one (continued from thread)..

bwaidelich commented:
yes, I like @aertmann s idea to use "fusion" as entry point and we already do that for the Neos UI with "ui"

mficzel commented:
I adjusted the option name to ``fusion.prototypeGeneratorClassName``, updated the documentation and added a Plugin Prototype Generator.

@aertmann this is somehow breaking but it will only go to 3.0 anyways.

dimaip commented:
@aertmann I think it would be better to make this breaking than to keep the two mechanisms. I doubt many people already use that feature anyways.

dimaip commented:
`prototypeGeneratorClassName` is very verbose though, any chances of making it shorter, e.g. `prototypeGenerator`?

mficzel commented:
Renamed the to ``fusion.prototypeGenerator``

aertmann commented:
looks great! nice one

dimaip commented:
Can we already merge this stuff into master now, or only after we branch off 3.0?
If it's possible to do it now, the PR needs to be rebased and I'll review it.

aertmann commented:
it's fine to merge breaking stuff for next major into master now

johannessteu commented:
Yeah, also branching will be done not soonish

mficzel commented:
I rebased the pr

mficzel commented:
I cannot reproduce the failing tests locally. Unit and functional tests are running fine. Also the error looks kind of unrelated to this change.

mficzel commented:
The errors were introduced in master with https://github.com/neos/neos-development-collection/commit/b2a1b0f6533a7d0373b19c636c543564b06a8a1d and are not caused by this pr. I will rebase again as soon as the master is green again.

mficzel commented:
Finally the tests are green again ...

kitsunet commented on TYPO3.TYPO3CR/Resources/Private/Schema/NodeTypes.schema.yaml:
```diff
@@ -75,6 +75,10 @@ additionalProperties:
     'options':
       type: dictionary
 
+    'fusion':
```
I was almost about to merge this, but this IMHO cannot happen, it implicitly binds CR to Fusion, which it shouldn't be. SORRY :/

The configuration should go under "options" and not be mentioned in this schema.

bwaidelich replied:
I agree to your point but TBH I don't think it's a better solution to "hide" this defacto dependency within some option that is not part of the schema.
IMO a good solution would be: extensible schemas. That would allow the Neos or Fusion package to extend the schema.
It shouldn't be to hard to adjust the `SchemaValidator` accordingly. In fact, it might already be able to do that.

kitsunet replied:
Right, but I thought `options` was the current way to make this schema extensible?
So Neos can declare the fusion subkey in options and all currently possible subkeys of fusion.

mficzel replied:
The key was renamed to ``options.fusion.prototypeGenerator`` as requested

kitsunet commented on TYPO3.Neos/Documentation/CreatingASite/NodeTypes/NodeTypeDefinition.rst:
```diff
@@ -119,6 +119,32 @@ The following options are allowed:
     Alternatively the class of a node label generator implementing
     ``TYPO3\TYPO3CR\Domain\Model\NodeLabelGeneratorInterface`` can be specified as a nested option.
 
+``fusion``
```
The below mentioned request to change the way the configuration is setup obviously makes a change here necessary.

mficzel commented:
Ranamed the configuration property to ``options.fusion.prototypeGenerator`` as requested by @kitsunet and rebased on current master  ... no clue why travis complains on postgres now ... will look into this later

kitsunet commented:
Thanks! Looks great. The Travis failure is not related, it's the timezone issue fixed by this PR (which is not yet upmerged) https://github.com/neos/neos-development-collection/pull/1214

mficzel commented:
The failing travis test looks like a wintertime-issue to me and seems unrelated to this pr.

mficzel commented:
I rebased on current master to get rid of the ci-failures. No changes other than that.
